### PR TITLE
Factor out kafka async wrapper code

### PR DESCRIPTION
**Pull request by ottomata**
This code has been moved to a separate module:
  @wikimedia/node-rdkafka-factory

and added here as an optional dependency.  The default-eventgate
factory will only attempt to require this if kafka.conf is set.

By factoring out this code, custom EventGate implementations
can require is separately from EventGate to create their own
Kafka produce functions.

Bug: T266058

### CODE REVIEW

**Comment by ottomata** [comment]
@wikimedia/node-rdkafka-factory initial commit for review is here: https://gerrit.wikimedia.org/r/c/node-rdkafka-factory/+/637026

We need to merge and publish that before this is merged.

File: README.md [hunk 1]
```diff
@@ -191,11 +192,14 @@ To use a custom EventGate implementation, set `eventgate_factory_module` to your
...
 
-## Default Kafka EventGate configuration
+## Default EventGate configuration
 
 The following values in `conf` will be used to instantiate a a default EventGate
 that extracts JSONSchemas from schema URIs, validates events using those
-schemas, and then produces them to Kafka.
+schemas, and then produces them to an output file and or Kafka.
+If `kafka.conf` is set, than Kafka will be used.  Since node-rdkafka-factory
+is an optional dependency, please make sure it (and node-rdkafka) are properly
```
Comment (Pchelolo) [inline README.md]: s/are/is

File: config.yaml [hunk 2]
```diff
@@ -71,16 +73,27 @@ services:
...
-        conf:
-          metadata.broker.list: 'localhost:9092'
-        # kafka topic conf goes here
-        topic_conf: {}
+      # If kafka.conf is defined, make sure the optionalDependency
+      # node-rdkafka-factory (and node-rdkafka) is installed properly.
+      # kafka:
+      #  conf:
+      #    metadata.broker.list: 'localhost:9092'
+      #    # If this is set, node-rdkafka-statsd will be used
```
Comment (Pchelolo) [inline config.yaml]: maybe no need to add a comment within a comment ( e.g. multitude of # symbols)

File: config.yaml [hunk 3]
```diff
@@ -71,16 +73,27 @@ services:
...
+      #    # to emit metrics to statsd via service-runners metrics
+      #    # object. Make sure the node-rdkafka-statsd optionalDependency
+      #    # is installed properly.
```
Comment (Pchelolo) [inline config.yaml]: maybe just 'is installed'? How can it be improperly installed? :)

File: lib/factories/default-eventgate.js [hunk 4]
```diff
@@ -168,34 +183,135 @@ function makeValidate(options, logger) {
...
+ *      node-rdkafka KafkaProducer topic configuration
+ * @param {Object} logger
+ * @param {Object} metrics
+ *      service-runner metrics object, used if statistics.interval.ms is set
+ *      to collect rdkafka metrics.
+ * @return {Promise<{EventGate~produce}>} (event, context) => Promise<event>>
  */
-function makeProduce(options, producer) {
+async function makeKafkaProduce(options, logger, metrics) {
+    const kafka = require('node-rdkafka-factory');
```
Comment (Pchelolo) [inline lib/factories/default-eventgate.js:234]: maybe wrap in try-catch and re-throw with a nice error message and a hint to install the dependency?

File: lib/factories/default-eventgate.js [hunk 5]
```diff
@@ -168,34 +183,135 @@ function makeValidate(options, logger) {
...
+        kafkaProduce = await makeKafkaProduce(options, logger, metrics);
+        logger.info('Producing valid events to Kafka at ' + options.kafka.conf['metadata.broker.list']);
+    }
+
+    // Return a new function that takes a single event argument for produce.
+    return async (event, context = {}) => {
+        if (fileProduce) {
+            await fileProduce(event, context);
+            await kafkaProduce(event, context);
+        }
```
Comment (Pchelolo) [inline lib/factories/default-eventgate.js:316]: This is clearly quite wrong.. what if we are only producing to a file? there should be 2 ifs

**Review by Pchelolo — CHANGES_REQUESTED** [review]

File: config.yaml (hunk 2 above)
Comment (ottomata) [reply config.yaml]: I reformatted a bit and left a couple in.  Its nice to be able to uncomment the block and still have a few comments in there.

File: lib/factories/default-eventgate.js (hunk 5 above)
Comment (ottomata) [reply lib/factories/default-eventgate.js:316]: OO good catch thank youl.

File: lib/factories/default-eventgate.js (hunk 4 above)
Comment (ottomata) [reply lib/factories/default-eventgate.js:234]: I think in this case the Error that is thrown is decent (and standard) when a dependency is not installed.

**Pull request merged**